common/persistence/elasticsearch/esProcessor.go

```diff
@@ -197,42 +196,79 @@ func (p *esProcessorImpl) bulkAfterAction(_ int64, requests []elastic.BulkableRe
 	if err != nil {
 		// This happens after configured retry, which means something bad happens on cluster or index
 		// When cluster back to live, processor will re-commit those failure requests
-		p.logger.Error("Error commit bulk request.", tag.Error(err))
 
+		isRetryable := isRetryableError(err)
+		p.logger.Error("Unable to commit bulk ES request.", tag.Error(err), tag.Bool(isRetryable))
 		for _, request := range requests {
 			p.logger.Error("ES request failed.", tag.ESRequest(request.String()))
 			p.metricsClient.IncCounter(metrics.ElasticSearchVisibility, metrics.ESBulkProcessorFailures)
+
+			if !isRetryable {
+				visibilityTaskKey := p.extractVisibilityTaskKey(request)
+				if visibilityTaskKey == "" {
+					continue
+				}
+				p.sendToAckChan(visibilityTaskKey, false)
+			}
 		}
 		return
 	}
 
+	responseIndex := p.buildResponseIndex(response)
 	for i, request := range requests {
-		visibilityTaskKey := p.getVisibilityTaskKey(request)
+		visibilityTaskKey := p.extractVisibilityTaskKey(request)
 		if visibilityTaskKey == "" {
 			continue
 		}
-		responseItem := response.Items[i]
-		for _, resp := range responseItem {
+
+		docID := p.extractDocID(request)
+		if docID == "" {
+			p.sendToAckChan(visibilityTaskKey, false)
```

Comment:
can we either 1) log here as well (just as you are logging below) or 2) can we remove this block and rely on the code on line 231?

Is it worth adding any metrics to cover some of these error paths (just for our own insight or in case we want to alert)?

Reply:
also, in the old code, if this was empty, we would ACK the message. Now we are going to NACK it (used to be true for this case, now its false). No implications here, right?

Reply:
1. Removed this block.
2. `extractDocID` writes to log and metrics itself.
3. Actually this should never happen. `_id` is always there.